parsers/typespec.c

```diff
@@ -0,0 +1,749 @@
+			else
+				c = next;
+		}
+	} while (c != EOF && c != '\n' && c != '\r');
+	return c;
+}
+
+static int skipMultiComment (void)
+{
+	int c = getcFromInputFile ();
+	int next = getcFromInputFile ();
+	
+	while (c != EOF && !(c == '*' && next == '/'))
+	{
+		c = next;
+		next = getcFromInputFile ();
+	}
+	
+	return getcFromInputFile ();
+}
+
+static void parseString (vString *const string, const int delimiter)
+{
+	while (true)
+	{
+		int c = getcFromInputFile ();
+
+		if (c == '\\' && (c = getcFromInputFile ()) != EOF)
+			vStringPut (string, c);
+		else if (c == EOF || c == delimiter)
+			break;
+		else
+			vStringPut (string, c);
+	}
+}
+
+static void parseIdentifier (vString *const string, const int firstChar)
+{
+	int c = firstChar;
+	do
+	{
+		vStringPut (string, c);
+		c = getcFromInputFile ();
+	} while (isIdentChar (c));
+	ungetcToInputFile (c);
+}
+
+static void readToken (tokenInfo *const token)
+{
+	int c;
+
+	token->type		= TOKEN_UNDEFINED;
+	vStringClear (token->string);
+
+getNextChar:
+
+	c = getcFromInputFile ();
+	c = skipWhitespace (c);
+
+	token->lineNumber   = getInputLineNumber ();
+	token->filePosition = getInputFilePosition ();
+
+	switch (c)
+	{
+		case EOF: token->type = TOKEN_EOF;					break;
+		case '(': token->type = TOKEN_OPEN_PAREN;			break;
+		case ')': token->type = TOKEN_CLOSE_PAREN;			break;
+		case ';': token->type = TOKEN_SEMICOLON;			break;
+		case ',': token->type = TOKEN_COMMA;				break;
+		case '.': 
+		{
+			/* Handle ... (spread) operator */
+			int c1 = getcFromInputFile ();
+			int c2 = getcFromInputFile ();
+			if (c1 == '.' && c2 == '.')
+			{
+				token->type = TOKEN_SPREAD;
+			}
+			else
+			{
+				ungetcToInputFile (c2);
+				ungetcToInputFile (c1);
+				token->type = TOKEN_PERIOD;
+			}
+			break;
+		}
+		case ':': token->type = TOKEN_COLON;				break;
+		case '{': token->type = TOKEN_OPEN_CURLY;			break;
+		case '}': token->type = TOKEN_CLOSE_CURLY;			break;
+		case '<': token->type = TOKEN_OPEN_ANGLE;			break;
+		case '>': token->type = TOKEN_CLOSE_ANGLE;			break;
+		case '=': token->type = TOKEN_EQUAL_SIGN;			break;
+		case '[': token->type = TOKEN_OPEN_SQUARE;			break;
+		case ']': token->type = TOKEN_CLOSE_SQUARE;			break;
+
+		case '\'':
+		case '"':
+			token->type = TOKEN_STRING;
+			parseString (token->string, c);
+			break;
+
+		case '/':
+		{
+			int d = getcFromInputFile ();
+			if (d == '/')  /* single-line comment */
+			{
+				skipSingleComment ();
+				goto getNextChar;
+			}
+			else if (d == '*')  /* multi-line comment */
+			{
+				skipMultiComment ();
+				goto getNextChar;
+			}
+			else
+			{
+				ungetcToInputFile (d);
+				token->type = TOKEN_UNDEFINED;
+			}
+			break;
+		}
+
+		case '@':  /* decorator */
+			token->type = TOKEN_DECORATOR;
+			c = getcFromInputFile ();
+			parseIdentifier (token->string, c);
+			break;
+
+		default:
+			if (! isIdentChar (c))
+				token->type = TOKEN_UNDEFINED;
+			else
+			{
+				parseIdentifier (token->string, c);
+				
```

Comment:
Remove tab chars at the beginning of lines.

Reply:
There are many unnecessary white spaces and tabs. I cannot remember how to configure, but git show can visualize them.

![image](https://github.com/user-attachments/assets/710c4988-9311-44b7-8c4f-93dab8e8353a)